### PR TITLE
WIP: MSC3588: Encrypted Stories As Rooms

### PR DESCRIPTION
**Pull request by krille-chan**
[Rendered](https://github.com/krillefear/matrix-doc/blob/main/proposals/3588-stories-as-rooms.md)

Signed-off-by: Christian Pauly christian-pauly@posteo.de

### CODE REVIEW

File: proposals/3588-stories-as-rooms.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+In this way, we make sure that every invitation to a new stories room actually contains at least one new post.
+
+Invitations to a new story room could be displayed as a normal new story in the UI while there should be a way to decline such an invitation, for example with a long press context menu.
+
+A post in a story room can then just be a normal `m.room.message` event of any message type, which then would be end-to-end encrypted.
+
+## Potential issues
+
+All posts in a story should disappear after 24 hours. In the future it might be easy to implement this using "Self destructive messages" <https://github.com/matrix-org/matrix-doc/pull/2228> but at the moment the only way we can do this, is to manage this on both sides:
```
Comment (Cadair) [inline proposals/3588-stories-as-rooms.md:82]: An alternative idea, would be to have a single "stories space" per user, where you can invite your contacts, and your contacts can join or leave to opt out of access to your stories, and then you can make each new story as a new room in that space using restricted join rules for access control.

From my very brief thinking about it it would mean you could kick everyone from the room after 24hours to "expire" the story, and then you can use `m.space.child` events in the "story space" as a way for people to decide if they want to see that specific story.

File: proposals/3588-stories-as-rooms.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
+# MSC3588: Stories As Rooms
+
+Stories are a feature mostly known from Instagram, WhatsApp, Facebook and SnapChat. They are a way to share text, images or videos with all of your friends at once while they have a limited lifetime of mostly 24 hours. In contrast to microblogging, stories are ephemeral and therefore users are motivated to speek freely.
```
Comment (uhoreg) [inline proposals/3588-stories-as-rooms.md:3]: Please wrap your lines

**Review by uhoreg — COMMENTED** [review]
Since this is marked as "WIP", I'm going to convert this to a draft PR.  Feel free to switch it back to non-draft when you think it's ready.

File: proposals/3588-stories-as-rooms.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+We have some requirements for this feature here:
+
+* Posts in a story should disappear after a certain lifetime
+* The user should be able to configure who can see their story
+* The user should be able to opt-out of receiving stories from specific users (without the need to fully ignore such users)
+* Posts in a story should be end-to-end encrypted (and therefore not be public)
+* Stories are **not** public. Users need to have some kind of relationship to see the each others stories
```
Comment (MTRNord) [inline proposals/3588-stories-as-rooms.md:11]: From a user perspective, I actually would like to have the ability to have it shared to anyone who knows me. Though that is likely pretty impossible without profiles as rooms to have some kind of "hey, here is the user's story".

File: proposals/3588-stories-as-rooms.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+### Stories Room
+
+Before a user posts their first story, they need to create a story room. This room should be marked with a creation type of `msc3588.stories.stories-room`. The stories room should not be created before the user actually wants to post their first story.
+
+The room should be private, encrypted and read only for all other users. While creating this room, the user should already invite all "contacts" they would like to share their story with. Apps should display a list of possible contacts to choose, which could be just a list of all users, the user shares a DM room with.
+
+If a user has less than 30 DM rooms, it should be fine that the client just preselects all of them but the user must be informed that they is going to invite those 30 people to a new room and that these people will be able to see each other's MXID which might not be wanted. All contacts which should **not** be invited to the room, should be added to a block list in a private account data object.
```
Comment (MTRNord) [inline proposals/3588-stories-as-rooms.md:28]: Why have an additional Blocklist if the room is already invite only?

File: proposals/3588-stories-as-rooms.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    }
+  ],
+  "invite": [
+    "@alice:example.com",
+    "@bob:example.com"
+  ]
+}
+```
+
+#### Example of the block list account data object:
```
Comment (MTRNord) [inline proposals/3588-stories-as-rooms.md:57]: Hm, is it possible to use the existing blocklist format that also mjolnir uses (If my other blocklist question comes to the resolution of "yes, we really need it")?

File: proposals/3588-stories-as-rooms.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+Therefore, it **will** be possible that stories are visible after 24 hours if one or both parts are not following these rules. However, it will always be possible to capture screenshots or to copy contents before they get redacted, regardless of what we try to prevent this. The automatic disappearing of stories is **not** a privacy feature. It just assumes that stories, older than a specific amount of time, are no longer relevant to the users.
+
+
+## Alternatives
+
+Profiles as rooms can be used to post content to a wider audiance.
+
+
+## Security considerations
```
Comment (MTRNord) [inline proposals/3588-stories-as-rooms.md:95]: Probably worth adding that matrix does not protect against downloading media, and also it may at the end not be gone from the media repo (aka if you know the link you still can access it)

File: proposals/3588-stories-as-rooms.md (hunk 1 above)
Comment (krillefear) [reply proposals/3588-stories-as-rooms.md:82]: Not sure. That would be huge overhead and needs a lot of time to join all the rooms. If I want to see the 20 Posts of a story from a user I need to join 20 rooms then?

File: proposals/3588-stories-as-rooms.md (hunk 3 above)
Comment (krillefear) [reply proposals/3588-stories-as-rooms.md:11]: This would be a different use case I think. These stories are very private and always e2ee. So it would make sense to only allow people to invite to the story room the user shares a DM room with in the GUI. Also e2ee rooms shouldn't be too large. I would say around 50 followers are a good limit.

For what you say it sounds more like a "public story" like on Instagram. This could be solved in a follow-up MSC.

File: proposals/3588-stories-as-rooms.md (hunk 4 above)
Comment (krillefear) [reply proposals/3588-stories-as-rooms.md:28]: We need this to know which contacts we do not want to have in the room for the moment. Every time we create a new story, the GUI should display a list of users we might want to invite which are {Users we share a DM room with} - {Users already follow the story} - {Users in the block list}.

File: proposals/3588-stories-as-rooms.md (hunk 5 above)
Comment (krillefear) [reply proposals/3588-stories-as-rooms.md:57]: Maybe the wording blocklist is not perfect here. It has nothing to do with ban. It is just a "Not-invite-list".

File: proposals/3588-stories-as-rooms.md (hunk 6 above)
Comment (krillefear) [reply proposals/3588-stories-as-rooms.md:95]: Good idea :-)

File: proposals/3588-stories-as-rooms.md (hunk 4 above)
Comment (Alch-Emi) [inline proposals/3588-stories-as-rooms.md:28]: If I understand correctly, users are only added to the blocklist when the story sender doesn't want to include them in the story.  Is there a mechanism in place for when story recipient doesn't want to receive stories?

In particular, I'm imagining a scenario where one friend adds another friend each time they send a story, but the second friend keeps declining the invite.  I think this is definitely plausible, if the sending friend doesn't remember adding the second friend the first time, or doesn't understand that them showing up in the invite list again means they've declined.

Maybe the user shouldn't be prompted to invite people with m.leave memberships unless they knock again?

Comment (krillefear) [reply proposals/3588-stories-as-rooms.md:28]: I already ran into this with the test implementation. Currently it looks like this:

1. User A wants to publish a new story but has no story room yet
2. Client creates a new story room and displays a list of contacts
3. User A selects User B to invite
4. User B declined the invitation
5. User A wants to publish another story
6. Client displays a list of contacts, which are not in the story room but also not in the blocklist, so only User B
7. User A can select User B to try to invite them again or just not select them so User B comes to the blocklist as well
8. In case of User A invites User B again, user B might contact User A with: "Eyy! I dont want to see your stupid story!! Stop inviting me over and over again!"
9. In the other case, everything is fine

File: proposals/3588-stories-as-rooms.md (hunk 1 above)
Comment (aaronraimist) [reply proposals/3588-stories-as-rooms.md:82]: > From my very brief thinking about it it would mean you could kick everyone from the room after 24hours to "expire" the story

That doesn't really work because users can view the contents of rooms that they have left or been kicked from. (Historical rooms has been broken in Element for many years but it is part of Matrix)

File: proposals/3588-stories-as-rooms.md [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+  "type": "msc3588.stories.block-list",
+  "content": {
+    "users": [
+      "@charley:example.com"
+    ]
+  }
+}
+```
+
+Clients should display such rooms in a different manner. Every user should have only one stories room but can in theory have multiple of them. The user must never give admin permissions to another person and should be aware, that all users in this room in theory can see each other's MXID.
```
Comment (kevincox) [inline proposals/3588-stories-as-rooms.md:70]: I would say that multiple story rooms should be the expected state. I may want to share different stories with different people so I would need a room for each set.

File: proposals/3588-stories-as-rooms.md (hunk 1 above)
Comment (Mikaela) [inline proposals/3588-stories-as-rooms.md:82]: Currently this could be done by having a `m.room.retention` state event

```json
{
      "min_lifetime": 86400000,
      "max_lifetime": 86400000
}
```

but that requires everyone to be using Synapses and to have enabled it in Synapse config, so I agree that https://github.com/vector-im/element-meta/issues/82 is the way to go.

File: proposals/3588-stories-as-rooms.md (hunk 6 above)
Comment (Mikaela) [reply proposals/3588-stories-as-rooms.md:95]: Ref: https://github.com/matrix-org/synapse/issues/1263

**Comment by TheOPtimal** [comment]
Guys, I feel like this is a useless PR either way - these kinds of social features should be left up to the ActivityPub social media platforms, and implementing it in a chat* protocol is just dumb and completely useless.

*I know matrix is technically more than just for chat, but currently that's basically all its used for

**Comment by Mikaela** [comment]
I would like to point out that this feature originates from a chat app, SnapChat, and was also implemented by WhatsApp in 2017, which I understand FluffyChat (where Stories are implemented in the latest release candidate) is directly challenging

**Comment by MTRNord** [comment]
> Guys, I feel like this is a useless PR either way - these kinds of social features should be left up to the ActivityPub social media platforms, and implementing it in a chat* protocol is just dumb and completely useless.
> 
> *I know matrix is technically more than just for chat, but currently that's basically all its used for

(I am aware that this gets flagged as offtopic or even spam now, but I have to write this message.)

You are fundamentally wrong on this. We have by now at least 5 maybe 6 social networks on matrix, all with different goals. Not only that, but matrix IS NOT a IM messaging protocol. It is a Data delivery protocol. Does this proposal have privacy issues? Probably. Is that bad? No. Don't make this a IM protocol, and please look up what people use the protocol for before writing these messages. This is not only incredible rude for people working on different usecases, but it also causes people to get mentally sad and harmed. In my opinion, pointless comments like you just did should be forbidden under a COC, but that's not my decision to make.  

Also, on your addition you made: If you agree this is not chat only. Why then stop people being able to MAKE other use cases. You can't make other use cases if you work against that and then claim it is not a chat proto because that's (in your world) only used for chat.

**Comment by richvdh** [comment]
folks, please remember:
 * be civil in your discussions. Take a deep breath before replying to a comment that makes you angry, and then explain politely why disagree.
 * Use comments on the on the diff rather than the PR, so that related discussions are threaded together.

Since the last three comments have been made on the PR rather than the diff, I'm going to hide them.

File: proposals/3588-stories-as-rooms.md [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+## Proposal
+
+The solution in this document works fully client side and requires no changes on the server. It is intended to be an optional feature for Matrix clients who would like to implement this feature, while it stays compatible with all other clients as well. Stories rooms will just be displayed as normal read-only rooms for them.
+
+### Stories Room
+
+Before a user posts their first story, they need to create a story room. This room should be marked with a creation type of `msc3588.stories.stories-room`. The stories room should not be created before the user actually wants to post their first story.
+
+The room should be private, encrypted and read only for all other users. While creating this room, the user should already invite all "contacts" they would like to share their story with. Apps should display a list of possible contacts to choose, which could be just a list of all users, the user shares a DM room with.
```
Comment (turt2live) [inline proposals/3588-stories-as-rooms.md:26]: The limitation of already sharing a DM/private room with the user should be enforced by the protocol, I think. At least until we have contact lists and we can enforce it properly.

**Comment by JokerGermany** [comment]
I don't like that you have to invite people. I think this is obtrusive...
There has to be another way...

File: proposals/3588-stories-as-rooms.md (hunk 3 above)
Comment (JokerGermany) [reply proposals/3588-stories-as-rooms.md:11]: I don't like that you have to invite people. I think this is obtrusive...
There has to be another way...

Perhaps something like a user hava a own stories spaces and there the new story rooms are added without noticing the user. (I suppose for this many things have to be changed...)
But when the user look in the space he see it and perhaps when you click on the user there is a stories tab.

**Comment by tezlm** [comment]
Adding even more features seems like a great way to break client compatibility.

I get that other clients could simply not implement stories and have them show up as rooms, but there is no easy way to "opt out" of getting sent invites from people who use clients who support them. 

I'm not against using matrix for non-instant messaging applications, but there is not really a good way to separate use cases currently. Even if different sets of required/optional/don't use MSCs were made for different use cases, syncing would result in extra garbage data that someone who doesn't want to see what amounts to a microblogging feature that belongs in a more social-media oriented app.

Maybe you don't want to have to use a separate app to look at people's stories. In that case, clients will slowly succumb to feature creep and code complexity, becoming more prone to bugs and harder to maintain. Plus extra ui that users have to deal with. And like the first sentence in this comment, people simply can't really ignore features without things breaking. (e2ee is supposedly optional for all clients [according to the spec](https://spec.matrix.org/v1.2/client-server-api/#modules), but it really isnt due to how much stuff relies on encryption. Yeah, I know not supporting this pr won't break things as spectacularly, but it's simply the pr I chose to leave this comment on)

tl;dr: Matrix needs a better way to separate use cases before the number of added MSCs get out of hand.

File: proposals/3588-stories-as-rooms.md [hunk 9]
```diff
@@ -0,0 +1,107 @@
+# MSC3588: Stories As Rooms
```
Comment (MTRNord) [inline proposals/3588-stories-as-rooms.md:1]: (reply to @tezlm)

>  syncing would result in extra garbage data for someone who doesn't want to see what amounts to a microblogging feature that belongs in a more social-media oriented app.

This probably can be solved using a sync filter if someone makes filters by room types possible. Likely this is already part of sliding sync. I would personally consider this a non-issue.
Also, this data doesn't randomly land in your account. You have to join the room. So I am not sure what your concern here is? If a user doesn't want it, he should just not join the room.

> In that case, clients will slowly succumb to feature creep and code complexity, becoming more prone to bugs and harder to maintain.

I believe this up to the client devs to decide if they implement it or not. I don't think it causes feature creep therefor.


Also, @tezlm please use threads. It is part of the spec progress and based on other MSCs you can clearly see that non thread comments are just ignored.

Comment (tezlm) [reply proposals/3588-stories-as-rooms.md:1]: > If a user doesn't want it, he should just not join the room.

...and how would I opt out of getting story room invites altogether?

> This probably can be solved using a sync filter if someone makes filters by room types possible. Likely this is already part of sliding sync. I would personally consider this a non-issue.

I didn't realize this existed, but it sounds very nice.

> I believe this up to the client devs to decide if they implement it or not. I don't think it causes feature creep therefor.

My comment doesn't apply to *only* this specific pr, but I chose to comment on this because it could have the most visible impact on people. A few of my friends and I were confused what stories were at first, and in the end it seems like it was client incompatibility - stories were just shown as and mixed into normal rooms, with not much information as to what they were for.

The main thing I'm worried about is the client devs deciding not to implement something. Matrix clients would end up like web browsers - only a few support every feature, and whenever someone uses a client that doesnt support certain features, their friends/family/peers tell them to switch to one that supports said feature.

Looking back at it now, this pr was probably not the best br to leave my comment on (eg. threads/polls would have even more of this problem) but I'll go with it for now.

> please use threads.

Sorry, didn't realize they were a thing.

File: proposals/3588-stories-as-rooms.md (hunk 8 above)
Comment (theotheroracle) [reply proposals/3588-stories-as-rooms.md:26]: wait, why ? why should i not be able to invite people i don't have marked as dms with ?

**Comment by krillefear** [comment]
This has been implemented in minesTrix and FluffyChat and will soon be part of the Matrix Dart SDK so we can remove the "needs implementation" label

**Comment by uhoreg** [comment]
Can you link to the implementations (preferably a PR or MR link) in the top comment of this PR so that someone can validate that the implementation matches the proposal?  Also, this MSC is still marked as a WIP -- I don't know if it makes sense to drop the "needs implementation" label from something that is still a WIP and may change in the future.  It may depend an what part of the MSC is still a WIP.

**Comment by mewtrino** [comment]
Why is this marked as abandoned when two clients implement it?

**Comment by uhoreg** [comment]
> Why is this marked as abandoned when two clients implement it?

The author has closed it.  Anyone is free to re-propose it, but for now, nobody is doing the work to get it into the spec.

**Comment by vanyasem** [comment]
@mewtrino the support for stories has been removed from FluffyChat as of [1.17.0](https://github.com/krille-chan/fluffychat/releases/tag/v1.17.0)

Stories were replaced with presence status messages in this pull request: https://github.com/krille-chan/fluffychat/pull/736
